File: PlagiarismCheckerPython/HashFunction.py
class HashFunction:

    def __init__(self, k_gram_length):
        self.base = 3
        self.prime = 1301081

        if k_gram_length > 2:
            self.base_offset = self.base % self.prime

            for i in range(1, k_gram_length - 1):
                self.base_offset = (self.base_offset * self.base) % self.prime

        else:
            self.base_offset = self.base



    def Hash(self, k_gram):

        hash = 0

        for i in range(len(k_gram)):
            hash = ((hash + ord(k_gram[i])) % self.prime) * self.base

        return hash / self.base



    def RollingHash(self, old_hash, prev_k_gram, k_gram):

        if prev_k_gram == k_gram:
            return self.Hash(k_gram)

        a = old_hash + self.prime
        b = ord(prev_k_gram[0]) * self.base_offset % self.prime
        c = (a - b) * self.base
        d = c + ord(k_gram[len(k_gram) - 1])
        return d % self.prime

File: PlagiarismChecker/HashFunction.js

module.exports = class HashFunction {

    constructor(kGramLength) {
        this.base = 3;
        this.prime = 1301081;

        if(kGramLength > 2) {
            this.baseOffset = this.base % this.prime;
            for (let i = 1; i < kGramLength - 1; i++) {
                this.baseOffset = (this.baseOffset * this.base) % this.prime;
            }

            //this.baseOffset *= this.base;
        } else
            this.baseOffset = this.base;

        //console.log('offset: ' + this.baseOffset);

    }


    Hash(kGram) {

        let hash = 0;

        for (let i = 0; i < kGram.length; i++) {
            hash = ((hash + kGram.charCodeAt(i)) % this.prime) * this.base;
        }

        return hash / this.base;
    }


    RollingHash(oldHash, prevKGram, kGram){

        if(prevKGram === kGram)
            return this.Hash(kGram);

        if(kGram === 'rnmentthathadkeptfaithwit')
            console.log('');

        var a = oldHash + this.prime;
        var b = prevKGram.charCodeAt(0) * this.baseOffset % this.prime;
        var c = (a - b) * this.base;
        var d = c + kGram.charCodeAt(kGram.length - 1);
        var e = d % this.prime;
        return e;

        /*
        return ((oldHash + this.prime - (prevKGram.charCodeAt(0) * this.baseOffset) % this.prime) *
            this.base + kGram.charCodeAt(kGram.length - 1)) % this.prime;
            */
    }
};
